Add Conformance.Imp: imptests with conformance

aniketd opened the pull request:
# Description

With a new `iteExpectLedgerRuleConformance` field added to `ImpTestEnv`,
we get an overridable function that can be executed within `trySubmitTx`
for every submitted transaction. To override this hook we have
`modifyImpInitExpectLedgerRuleConformance` the function. We add a new
`Conformance.Imp` module to `cardano-ledger-conformance` package, to
import `Conway.Imp` and run those tests with a modified "hook" that
runs the `LEDGER` rule from Agda on the Tx and `checkConformance` on
the results.

Additions:
- `iteExpectLedgerRuleConformance` field to `ImpTestEnv`.
- `modifyImpInitExpectLedgerRuleConformance` function to override the
hook.
- Add lenses for all fields of `LedgerEnv`.

Changes:
- Change `trySubmitTx` to run the `iteExpectLedgerRuleConformance` for
every submitted Tx.

### The Problem

We get a failure from the Agda side, telling us that the `witsVKeyNeeded utxo txb` is not a subset of `witsKeyHashes` as it should be. We have isolated the issue to be with the translation of `spendableInputs`, since it is most likely that those aren't being handled in a manner that fits into the problem we are trying to solve with this conformance.

This PR can be merged as it is, since @Soupstraw and I want to work in parallel on related issues.

Closes #4725 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

lehins commented:
@aniketd #4745 was merged yesterday. You can rebase this one on master and finish it up

Soupstraw commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -302,6 +304,7 @@ instance ShelleyEraImp era => ImpSpec (LedgerSpec era) where
             ImpTestEnv
               { iteFixup = fixupTx
               , iteCborRoundTripFailures = True
+              , iteHook = \_ _ _ _ -> pure ()
```
I think we should be more specific about where this hook gets called, so a better name would be something like `iteSubmitTxHook`

lehins replied:
Totally agree with the suggestion of renaming. I would even go further and suggest avoiding using name `Hook`. We only used it in passing to describe this concept. A better name would be: `iteExpectLedgerRuleConformance`

Soupstraw reviewed (approved):
LGTM

teodanciu commented on libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp.hs:
```diff
@@ -0,0 +1,144 @@
+spec :: Spec
+spec =
+  withImpInit @(LedgerSpec Conway) $ do
+    xdescribe "Tx conformance"
```
So they are not passing yet?  The plan is to merge and fix them later?

Soupstraw replied:
Yeah, that's how I usually did with conformance tests also, this PR just sets up the infrastructure for running the tests. It's hard to predict how long it might take to fix a conformance failure + there might be other issues cropping up after one is fixed. I'd like to have this PR merged so I can add a similar hook to `passTick` without having to worry about rebasing.

teodanciu replied:
Makes sense, thanks for explaining.

Soupstraw commented on libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp.hs:
```diff
@@ -0,0 +1,144 @@
+          bimapM
+            (traverse toTestRep)
+            (toTestRep . inject @_ @(ExecState ConwayFn "LEDGER" era) . fst)
+            impRuleResult
```
This can be replaced with `runConformance`

aniketd replied:
Unfortunately it cannot be replaced, since it also, in addition, calls `tryRunImpRule` which we already do in `trySubmitTx`. I could refactor `runConformance` but that would mean we need to change the name too.

Soupstraw commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -767,13 +784,34 @@ impWitsVKeyNeeded txBody = do
+      LedgerEnv era ->
+      LedgerState era ->
+      Tx era ->
+      ImpTestM era ()
```
I think the environment and state arguments are unnecessary, because you already have access to those via `getsNES`

Soupstraw replied:
And the result of running the rule can also be computed inside the hook, but I guess it's faster to use the precomputed result, so it's fine to keep that as an argument

lehins replied:
By the time this hook is called the state has already been modified.
So, we do need to pass the result and the state to this hook. Environment on the other hand is not necessary, however it is best to pass it as well for consistency.

Moreover, I am leaning towards making it an `Expectation`. This would limit access of this hook to the ImpTest state and protect against mistakes that would affect the outcome of the imp test itself.

aniketd replied:
We need to `translateInputs` within the `iteExpectLedgerRuleConformance` since the translation functionality lives in `cardano-ledger-conformance`. Making the return type of the hook to `Expectation`, my guess is it would require a `liftIO` and the function obtained from `withRunInIO` to be passed to the hook. Please correct me if I'm wrong. I thought it will be a little cumbersome because within `trySubmitTx` we would need to change the context from `ImpTestM era ()` to `IO ()` with `liftIO` to call the hook and within the hook, we would need to change the context back to `ImpTestM (LedgerSpec era) ()` with `runInIO` to call `translateInputs`?

Soupstraw replied:
You should just be able to translate all the inputs using `runSpecTransM`, that's what `translateInputs` is doing internally.

aniketd replied:
I have pushed the update so this is now again ready for review. :raised_hands: Thanks!

lehins commented on libs/ImpSpec/src/Test/ImpSpec/Internal.hs:
```diff
@@ -97,6 +98,12 @@ deriving instance (Eq (ImpSpecEnv t), Eq (ImpSpecState t)) => Eq (ImpInit t)
+impInitEnvL = lens impInitEnv $ \x y -> x {impInitEnv = y}
+
+impInitStateL :: Lens' (ImpInit t) (ImpSpecState t)
+impInitStateL = lens impInitState $ \x y -> x {impInitState = y}
```
We are not adding any lenses to ImpSpec because that would result in an unnecessary dependency in a reusable package.

Moreover, we don't need to add a lens every time we have a record type. If a lens is going to be used only once (like in this PR) it is ok to just use regular record syntax.

lehins commented on libs/ImpSpec/ImpSpec.cabal:
```diff
@@ -47,7 +47,8 @@ library
         text,
-        unliftio
+        unliftio,
+        microlens
```
Please revert any changes to `ImpSpec`. It has already been extracted into [its own repo](https://github.com/input-output-hk/ImpSpec) and will be removed from `cardano-ledger` very soon.

lehins replied:
See #4763

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -618,16 +621,30 @@ modifyImpInitProtVer ::
+        .~ ProtVer ver 0
+    )
+
+modifyImpInitHook ::
```
Suggestion about the name applies to this modification function as well

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -1551,6 +1593,20 @@ withPostFixup ::
+  ImpTestM era a ->
+  ImpTestM era a
+withHook f = local $ iteHookL .~ f
+
```
Unused. Please remove
```suggestion
```

lehins commented on libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp.hs:
```diff
@@ -0,0 +1,144 @@
+spec =
+  withImpInit @(LedgerSpec Conway) $ do
+    xdescribe "Tx conformance"
+      . modifyImpInitProtVer @Conway (natVersion @10)
```
These modifiers can be moved higher in the call right after `withImpInit`, since we only do conformance in PV 10 and the hook is the same for all tests in here

lehins reviewed (approved):
Looks good, but a few minor changes are still needed before we can merge it

aniketd committed:
Add Conformance.Imp: ImpTests with conformance.

With a new `iteExpectLedgerRuleConformance` field added to `ImpTestEnv`,
we get an overridable function that can be executed within `trySubmitTx`
for every submitted transaction. To override this hook we have
`modifyImpInitExpectLedgerRuleConformance` the function. We add a new
`Conformance.Imp` module to `cardano-ledger-conformance` package, to
import `Conway.Imp` and run those tests with a modified "hook" that
runs the `LEDGER` rule from Agda on the Tx and `checkConformance` on
the results.

Additions:
- `iteExpectLedgerRuleConformance` field to `ImpTestEnv`.
- `modifyImpInitExpectLedgerRuleConformance` function to override the
hook.
- Add lenses for all fields of `LedgerEnv`.

Changes:
- Change `trySubmitTx` to run the `iteExpectLedgerRuleConformance` for
every submitted Tx.

lehins commented on libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp.hs:
```diff
@@ -0,0 +1,115 @@
+                }
+          }
+  -- translate inputs
+  let failOnLeft = either (assertFailure . unpack) pure
+  (specEnv, specState, specSignal) <-
+    (,,)
+      <$> failOnLeft (runSpecTransM ctx $ toSpecRep env)
+      <*> failOnLeft (runSpecTransM ctx $ toSpecRep state)
+      <*> failOnLeft (runSpecTransM ctx $ toSpecRep signal)
```
We already have an expectation like that in `ImpSpec`: [`expectRight`](https://hackage.haskell.org/package/ImpSpec-0.1.0.0/docs/Test-ImpSpec-Expectations.html#v:expectRight)
```suggestion
  (specEnv, specState, specSignal) <-
    (,,)
      <$> expectRight (runSpecTransM ctx $ toSpecRep env)
      <*> expectRight (runSpecTransM ctx $ toSpecRep state)
      <*> expectRight (runSpecTransM ctx $ toSpecRep signal)
```

Not sure if the contents of `Left` have a `Show` instance, cause that prevent `expectRight` usage. If that is the case then I'd suggest adding `HasCallStack` constraint on `failOnLeft`, so it would be possible to debug which one has failed.

In any case, this is a minor suggestion, so I am not gonna block this PR because of it. Feel free to apply this suggestion in the follow up work.

lehins commented on libs/cardano-ledger-conformance/test/Test/Cardano/Ledger/Conformance/Imp.hs:
```diff
@@ -0,0 +1,115 @@
+        xdescribe "Tx conformance" $ it "Tx conformance" $ do
+          _ <- submitConstitution @Conway SNothing
+          passNEpochs 2
+        xdescribe "Test.Cardano.Ledger.Conway.Imp conformance" $ ConwayImp.conwaySpec @Conway
```
Module names are not very useful in the test names.
```suggestion
        xdescribe "Conway Imp conformance" $ ConwayImp.conwaySpec @Conway
```

lehins reviewed (approved):
Great work. I've made couple suggestions, but neither one is a blocking suggestion, so in order not to block the follow up work I am gonna resolve them and merge the PR.
